### ISSUE TITLE
Fan oscillate switch state not remembered when fan switches off/on, no Siri control

### ISSUE THREAD

**Issue by SteveIsles** [joe-ng/homebridge-dyson-link#20]
The oscillate switch as part of the main fan control (no longer a "Fan Rotation" separate switch) doesn't allow for Siri control as far as I can tell: no formulation of "fan oscillation" "oscillate the fan in the living room" etc. works.  Is there a way to rename or register oscillate so it is controllable by Siri?

Slightly different but related, the oscillate switch's state is not respected when a fan is turned on, so that you need to toggle it off and on for it to stay in the oscillate state.

Basically, I always have to open the Home app each time I turn on the fan to flip it on and off to enable fan rotation, whereas before as a separate switch I could write an automation to change the rotation state.

I think if the "respect last state" issue is fixed (perhaps by "forcing" homebridge to toggle the fan rotation state off and then on when powering on the fan) the need to enter the app each time will be diminished.

**Comment by SteveIsles** [comment]
@joe-ng any update on this?  Right now, there's no way to turn the fan on and it remember the oscillation state or state of night mode.  If you have the night mode switch enabled _before_ turning on the fan (e.g. increase fan speed from 0%) the night mode switch will stay on, but it will not _actually_ be on.  You have to toggle the switch off and on for it to take effect.

Same for oscillate.  I think an elegant fix would be for it to send the command to the fan to turn those on each time the fan speed is set to a non-zero value, so that when a fan is turned on, oscillation and night-mode state etc. is active from the beginning...

**Comment by joe-ng** [comment]
Yup, I tried to make a patch for this in 1.8.2 and you may try to see if that works for you

**Comment by SteveIsles** [comment]
So, I just installed 1.8.2, and I see no improvement in behavior.  If I turn on oscillate first, then adjust then turn the fan speed on, the fan will not oscillate.  I can then turn the oscillate off and on and the fan begins to rotate.  Same behavior if I'm talking about the fan's night mode switch.  Did you see a change of behavior on your end?  I was running your 1.8.1 this morning and now 1.8.2 after it is out.

Interestingly I found a workaround of sorts for turning the fan oscillation on via Siri: you can create a scene with the Fan with a non-zero speed, the auto vs. manual the way you want, and oscillate on.  Activating the scene will turn on the fan(s) and deactivating the scene will turn them off.

Of course, a down-side of this approach is that it changes the speed to some preset value, but you can easily use Siri after activating the "fan oscillate" scene to tell it to "set the fan speed to xx%" etc.

**Comment by SteveIsles** [comment]
*UPDATE:* Apparently I was still running 1.8.1 — I rebooted and now I see different behavior, but unfortunately differently is bad.  If I turn on oscillate first and then try to turn on the fan, I get a "No Response" in the Home app (error status) and the fan does not turn on.  If the fan is turned on first, then turning on oscillation afterwards also triggers "no response."  I restarted homebridge 2x and saw the same behavior, and thus installed 1.8.1 again and it seems to be working.  So something with the change in 1.8.2 puts the fan in a bad state with the new behavior.

**Comment by joe-ng** [comment]
On my side, i can, say, turn on the fan and then turn on the oscillation and after that, if i turn off the fan and turn it back on, the oscillation will be stored and the log would be liked below

[DysonPlatform] Dyson - Set State:{"fmod":"FAN"}
[DysonPlatform] Dyson Try to restore the fan speed state from home app to 80
[DysonPlatform] Dyson - Set State:{"fnsp":8}
[DysonPlatform] Dyson Try to restore the fan swing state from home app
[DysonPlatform] Dyson - Set State:{"oson":"ON"}

can you send me the log you had when you encountered the issue?

**Comment by joe-ng** [comment]
thanks, just made a 1.8.3 to fix that

**Comment by SteveIsles** [comment]
This sees to behave better @joe-ng, but there are some weird issues/bugs.  I have 2 of the Dyson Cool tall purifiers and one of the Dyson hot/cold short purifiers.  The Cool fans now have the auto switch back (I believe you did this change in 1.8) — however, the auto/manual switch in the Cool fan "details" screen is still there.  It seems that the switch's status is "True" but the manual/auto "details page" switch is not actively used.  Moreover, the Dyson Hot/Cool short purifier does not have the same "Auto - Fan" switch and instead only has the manual/auto details page switch.

This is related to this issue because the UI is different for the two similar devices, and there are redundant switches whereas only 1 property seems to be used/updated.  If I switch from "Auto to Manual" on the details page, the fan will turn on in Auto mode anyway...

**Comment by Krocko** [comment]
With v2.0.1 i get the following error:

```
Apr 29 10:51:51 raspberrypi homebridge[13357]: /usr/local/lib/node_modules/homebridge-dyson-link/DysonLinkAccessory.js:168
Apr 29 10:51:51 raspberrypi homebridge[13357]: this.fan.removeCharacteristic(TargetFanState);
Apr 29 10:51:51 raspberrypi homebridge[13357]: ^
Apr 29 10:51:51 raspberrypi homebridge[13357]: ReferenceError: TargetFanState is not defined
Apr 29 10:51:51 raspberrypi homebridge[13357]: at DysonLinkAccessory.initFanState (/usr/local/lib/node_modules/homebridge-dyson-link/DysonLinkAccessory.js:168:47)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at new DysonLinkAccessory (/usr/local/lib/node_modules/homebridge-dyson-link/DysonLinkAccessory.js:26:14)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at config.accessories.forEach (/usr/local/lib/node_modules/homebridge-dyson-link/index.js:74:37)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at Array.forEach (<anonymous>)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at getDevicesFromAccount (/usr/local/lib/node_modules/homebridge-dyson-link/index.js:50:49)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at DysonPlatform.getDevicesFromAccount (/usr/local/lib/node_modules/homebridge-dyson-link/index.js:107:13)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at API.DysonPlatform.api.on (/usr/local/lib/node_modules/homebridge-dyson-link/index.js:49:26)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at API.emit (events.js:185:15)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at Server.run (/usr/local/lib/node_modules/homebridge/lib/server.js:98:13)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at module.exports (/usr/local/lib/node_modules/homebridge/lib/cli.js:40:10)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at Module._compile (internal/modules/cjs/loader.js:654:30)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at Object.Module._extensions..js (internal/modules/cjs/loader.js:665:10)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at Module.load (internal/modules/cjs/loader.js:566:32)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at tryModuleLoad (internal/modules/cjs/loader.js:506:12)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at Function.Module._load (internal/modules/cjs/loader.js:498:3)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at Function.Module.runMain (internal/modules/cjs/loader.js:695:10)
Apr 29 10:51:51 raspberrypi homebridge[13357]: at startup (internal/bootstrap/node.js:201:19)
Apr 29 10:51:52 raspberrypi homebridge[13357]: at bootstrapNodeJSCore (internal/bootstrap/node.js:516:3)

```

**Comment by joe-ng** [comment]
ar, sorry, that was a typo. thanks for the feedback. just fixed that in 2.1.1

**Comment by SteveIsles** [comment]
This works great, now the Auto/Manual detail option is lost from all Cool Tower fans — but why not do the same for the Cool/Hot tower?  Wouldn’t that device have the same issue that required you to move the auto/manual detail control to be a separate switch?   Or is there something I’m missing...

**Comment by joe-ng** [comment]
okay, i moved that for cool/hot device too now.
